src/io/flutter/run/bazelTest/BazelTestConfigProducer.java

```diff
@@ -0,0 +1,136 @@
+import io.flutter.FlutterUtils;
+import io.flutter.bazel.Workspace;
+import io.flutter.dart.DartPlugin;
+import io.flutter.pub.PubRoot;
+import io.flutter.run.FlutterRunConfigurationProducer;
+import io.flutter.run.test.TestConfig;
+import io.flutter.run.test.TestConfigType;
+import io.flutter.run.test.TestConfigUtils;
+import io.flutter.run.test.TestFields;
+import io.flutter.sdk.FlutterSdk;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Determines when we can run a test using "flutter test".
+ */
+public class BazelTestConfigProducer extends RunConfigurationProducer<BazelTestConfig> {
+
+  protected BazelTestConfigProducer() {
+    super(FlutterBazelTestConfigurationType.getInstance());
+  }
+
+  private static boolean isFlutterContext(@NotNull ConfigurationContext context) {
+    final PsiElement location = context.getPsiLocation();
+    return location != null;
+  }
+
+  /**
+   * If the current file looks like a Flutter test, initializes the run config to run it.
+   * <p>
+   * Returns true if successfully set up.
+   */
+  @Override
+  protected boolean setupConfigurationFromContext(BazelTestConfig config, ConfigurationContext context, Ref<PsiElement> sourceElement) {
+    if (!isFlutterContext(context)) return false;
+
+    final PsiElement elt = context.getPsiLocation();
+    if (elt instanceof PsiDirectory) {
+      return setupForTarget(config, context, ((PsiDirectory)elt).getVirtualFile().getPath());
+    }
+
+    final DartFile file = FlutterRunConfigurationProducer.getDartFile(context);
+    if (file == null) {
+      return false;
+    }
+
+    final String testName = TestConfigUtils.findTestName(elt);
+    if (testName != null) {
+      return setupForSingleTest(config, context, file, testName);
+    }
+
+    return setupForDartFile(config, context, file);
+  }
+
+  private boolean supportsFiltering(@Nullable FlutterSdk sdk) {
+    return sdk != null && sdk.getVersion().flutterTestSupportsFiltering();
+  }
+
+  private boolean setupForSingleTest(BazelTestConfig config, ConfigurationContext context, DartFile file, String testName) {
+    final VirtualFile testFile = verifyFlutterTestFile(config, context, file);
+    if (testFile == null) return false;
+
+    config.setFields(BazelTestFields.forTestName(testName, testFile.getPath(), Workspace.load(context.getProject())));
+    config.setGeneratedName();
+
+    return true;
+  }
+
+  private boolean setupForDartFile(BazelTestConfig config, ConfigurationContext context, DartFile file) {
+    final VirtualFile testFile = verifyFlutterTestFile(config, context, file);
+    if (testFile == null) return false;
+
+    config.setFields(BazelTestFields.forFile(testFile.getPath(), Workspace.load(context.getProject())));
+    config.setGeneratedName();
+
+    return true;
+  }
+
+  private VirtualFile verifyFlutterTestFile(BazelTestConfig config, ConfigurationContext context, DartFile file) {
```

Comment:
Possible follow up: `@Nullable` and `@NotNull` propagation in IntelliJ can help prevents NPEs